src/Computation/SymbolicExpression.cs

```diff
@@ -0,0 +1,389 @@
+        }
+
+        public IExpression Xor(IExpression expression)
+        {
+            return Binary(expression, (l, r) => new Xor(l, r));
+        }
+
+        public IExpression ZeroExtend(Bits size)
+        {
+            return size > Size
+                ? Unary(v => new ZeroExtend(size, v))
+                : this;
+        }
+
+        private IExpression Unary(Func<IValue, IValue> func)
+        {
+            return new SymbolicExpression(_contextFactory, _collectionFactory,
+                func(Value), Constraints);
+        }
+
+        private IExpression Binary(IExpression y, Func<IValue, IValue, IValue> func)
+        {
+            return Binary((IValueExpression) y, func);
+        }
+
+        private IExpression Binary(IValueExpression y, Func<IValue, IValue, IValue> func)
+        {
+            return new SymbolicExpression(_contextFactory, _collectionFactory,
+                func(Value, y.Value), Constraints.Concat(y.Constraints).ToArray());
+        }
+
+        private IExpression Ternary(IExpression y, IExpression z, Func<IValue, IValue, IValue, IValue> func)
+        {
+            return Ternary((IValueExpression) y, (IValueExpression) z, func);
+        }
+
+        private IExpression Ternary(IValueExpression y, IValueExpression z, Func<IValue, IValue, IValue, IValue> func)
+        {
+            return new SymbolicExpression(_contextFactory, _collectionFactory,
+                func(Value, y.Value, z.Value), Constraints.Concat(y.Constraints.Concat(z.Constraints)).ToArray());
+        }
+
+        private IExpression Evaluate(IPersistentSpace space)
+        {
+            using var model = space.GetModel(Constraints);
+
+            return new ConstantExpression(_contextFactory, _collectionFactory,
+                ConstantUnsigned.Create(Value.Size, model.Evaluate(Value)));
+        }
+
+        private BigInteger AsConstant()
+        {
+            using var handle = _contextFactory.Create();
+
+            return Value is Float && Value.AsFloat(handle.Context).Simplify().IsFPNaN
+                ? Value.Size.GetNan(handle.Context)
+                : AsConstant(Value.AsBitVector(handle.Context).Simplify());
+        }
+
+        private static BigInteger AsConstant(Expr expr)
+        {
+            return expr.IsNumeral
+                ? ((BitVecNum) expr).BigInteger
+                : throw new IrreducibleSymbolicExpressionException();
+        }
+
+        public static SymbolicExpression Create(IContextFactory contextFactory, ICollectionFactory collectionFactory,
+            IValue value, IEnumerable<Func<IExpression, IExpression>> constraints)
+        {
+            var unconstrained = new SymbolicExpression(contextFactory, collectionFactory,
+                value, Array.Empty<IValue>());
+
+            return new SymbolicExpression(contextFactory, collectionFactory,
+                value, constraints
+                    .Select(c => ((SymbolicExpression) c(unconstrained)).Value)
```

Comment:
I meant to say previously. How do we know that this cast is always safe?

Reply:
That invariant is enforced within this class. If the input to the `Func` is a `SymbolicExpression` then the output is guaranteed to be too. However, I will actually change the cast to `IValueExpression` since that works just as well.